validator/db/kv/attestation_history_v2.go

```diff
@@ -340,3 +343,105 @@ func (store *Store) shouldMigrateAttestations() (bool, error) {
 }
+
+// HighestSignedSourceEpoch returns the highest signed source epoch for a validator public key.
+// If no data exists, returning 0 is a sensible default.
+func (store *Store) HighestSignedSourceEpoch(ctx context.Context, publicKey [48]byte) (uint64, error) {
+	ctx, span := trace.StartSpan(ctx, "Validator.HighestSignedSourceEpoch")
+	defer span.End()
+
+	var err error
+	var highestSignedSourceEpoch uint64
+	err = store.view(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(newHistoricAttestationsBucket)
+		valBucket := bucket.Bucket(publicKey[:])
+		if valBucket == nil {
+			return fmt.Errorf("validator history empty for public key: %#x", publicKey)
+		}
+		highestSignedSourceBytes := valBucket.Get(highestSignedSourceKey)
+		if len(highestSignedSourceBytes) == 0 {
+			return nil
+		}
+		highestSignedSourceEpoch = bytesutil.BytesToUint64BigEndian(highestSignedSourceBytes)
+		return nil
+	})
+	return highestSignedSourceEpoch, err
+}
+
+// HighestSignedTargetEpoch returns the highest signed target epoch for a validator public key.
+// If no data exists, returning 0 is a sensible default.
+func (store *Store) HighestSignedTargetEpoch(ctx context.Context, publicKey [48]byte) (uint64, error) {
+	ctx, span := trace.StartSpan(ctx, "Validator.HighestSignedTargetEpoch")
+	defer span.End()
+
+	var err error
+	var highestSignedTargetEpoch uint64
+	err = store.view(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(newHistoricAttestationsBucket)
+		valBucket := bucket.Bucket(publicKey[:])
+		if valBucket == nil {
+			return fmt.Errorf("validator history empty for public key: %#x", publicKey)
+		}
+		highestSignedTargetBytes := valBucket.Get(highestSignedTargetKey)
+		if len(highestSignedTargetBytes) == 0 {
+			return nil
+		}
+		highestSignedTargetEpoch = bytesutil.BytesToUint64BigEndian(highestSignedTargetBytes)
+		return nil
+	})
+	return highestSignedTargetEpoch, err
+}
+
+// SaveHighestSignedSourceEpoch saves the highest signed source epoch for a validator public key.
+func (store *Store) SaveHighestSignedSourceEpoch(ctx context.Context, publicKey [48]byte, epoch uint64) error {
+	ctx, span := trace.StartSpan(ctx, "Validator.SaveHighestSignedSourceEpoch")
+	defer span.End()
+
+	return store.update(func(tx *bolt.Tx) error {
+		bucket := tx.Bucket(newHistoricAttestationsBucket)
+		valBucket, err := bucket.CreateBucketIfNotExists(publicKey[:])
+		if err != nil {
+			return fmt.Errorf("could not create bucket for public key %#x", publicKey)
+		}
+
+		// If the incoming epoch is higher than the highest signed epoch, override.
+		highestSignedSourceBytes := valBucket.Get(highestSignedSourceKey)
+		var highestSignedSourceEpoch uint64
+		if len(highestSignedSourceBytes) != 0 {
+			highestSignedSourceEpoch = bytesutil.BytesToUint64BigEndian(highestSignedSourceBytes)
+		}
+		if len(highestSignedSourceBytes) == 0 || epoch > highestSignedSourceEpoch {
+			if err := valBucket.Put(highestSignedProposalKey, bytesutil.Uint64ToBytesBigEndian(epoch)); err != nil {
```

Comment:
Let's be very careful here. This isn't the highestSignedProposalKey, likely a copy paste mistake